### PR TITLE
Fix issue 20023 - Separate compilation breaks dip1000 / dip1008 @safety

### PR DESCRIPTION
**Pull request by dkorpel**
Opening to see how the test suite reacts to it

### CODE REVIEW

**Comment by thewilsonator** [comment]
```
Warning: struct HasNonConstToHash has method toHash, however it cannot be called with const(HasNonConstToHash) this.
  HasNonConstToHash.toHash defined here: src/test_hash.d(156)
../../src/core/internal/convert.d(829): Error: returning `(cast(const(ubyte)*)&val)[0..12]` escapes a reference to parameter `val`
../../src/core/internal/convert.d(829):        perhaps remove `scope` parameter annotation so `return` applies to `ref`
../../src/core/internal/convert.d(817): Error: template instance `core.internal.convert.toUbyte!(S1)` error instantiating
../../src/core/internal/hash.d-mixin-543(639):        instantiated from here: `toUbyte!(S2)`
src/test_hash.d(200):        instantiated from here: `hashOf!(S3)`
../../src/core/internal/convert.d(829): Error: returning `(cast(const(ubyte)*)&val)[0..16]` escapes a reference to parameter `val`
../../src/core/internal/convert.d(829):        perhaps remove `scope` parameter annotation so `return` applies to `ref`
../../src/core/internal/convert.d(829): Error: returning `(cast(const(ubyte)*)&val)[0..1]` escapes a reference to parameter `val`
../../src/core/internal/convert.d(829):        perhaps remove `scope` parameter annotation so `return` applies to `ref`
../../src/core/internal/hash.d-mixin-543(639): Error: template instance `core.internal.convert.toUbyte!(shared(U))` error instantiating
src/test_hash.d(243):        instantiated from here: `hashOf!(shared(V))`
```

**Comment by thewilsonator** [comment]
Dependant PR merged

**Comment by dkorpel** [comment]
How do you test separate compilation? By the looks of it, EXTRA_FILES are compiled in one go with the test.

**Comment by dkorpel** [comment]
The test fails on Windows because it prints `fail_compilation/extra-files` with a forward slash (as specified by `-I`) instead of a backslash. Other tests with `-Ifail_compilation/extra-files` don't seem to have this problem.

**Comment by MoonlightSentinel** [comment]
> How do you test separate compilation? By the looks of it, EXTRA_FILES are compiled in one go with the test.

Use `EXTRA_SOURCES` + `COMPILE_SEPARATELY` ([README](https://github.com/dlang/dmd/blob/master/test/README.md#test-configuration)).

> The test fails on Windows because it prints `fail_compilation/extra-files` with a forward slash (as specified by `-I`) instead of a backslash. Other tests with `-Ifail_compilation/extra-files` don't seem to have this problem.

Probably because those files are also passed on the command line and hence normalized internally (contrary to those found on an import path).

File: test/fail_compilation/test20023.d [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// REQUIRED_ARGS: -preview=dip1000 -preview=dip1008 -Ifail_compilation/extra-files
+// https://issues.dlang.org/show_bug.cgi?id=20023
+/*
+TEST_OUTPUT:
+---
+fail_compilation/extra-files/test20023b.d(8): Error: scope variable `e` may not be returned
+fail_compilation/test20023.d(15): Error: template instance `test20023b.threw!()` error instantiating
+---
+*/
+import test20023b;
```
Comment (MoonlightSentinel) [inline test/fail_compilation/test20023.d:10]: Move `test20023b.d` to `imports` and reference it as `imports.test20023b`.
Then remove the additional import path defined above.

File: src/dmd/escape.d [hunk 2]
```diff
@@ -988,17 +988,14 @@ bool checkThrowEscape(Scope* sc, Expression e, bool gag)
         if (v.isScope() && !v.iscatchvar)       // special case: allow catch var to be rethrown
                                                 // despite being `scope`
         {
-            if (sc._module && sc._module.isRoot())
+            // Only look for errors if in module listed on command line
```
Comment (MoonlightSentinel) [inline src/dmd/escape.d:991]: This comment no longer applies?

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**